### PR TITLE
Unify text-shadows with the blur filter.

### PR DESCRIPTION
**Pull request by glennw**
This removes any special code for text shadows. A text shadow is
now simply a Picture with text runs and/or line decorations that
has a blur filter applied to it.

This is possible due to the recent work related to removing the line
decoration and cached text run primitives.

This greatly simplifies the shadow code, and a lot of the functionality
of the text shadows (e.g. fast path shadows) just fall out of the
existing Picture functionality for blur filters.

Next step is to expand the primitive types that can run through the
shadow paths, which will then allow us to unify the drop-shadow
filter to be a simple shadow + blur filter.

NOTE: The PictureKind enum now has only a single kind (Image). I could
      have removed that in this PR, and added the fields to Picture
      directly. However, that will result in a lot of indentation
      changes. So I'll do that as a follow up PR to make this one
      easier to review.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/2556)

### CODE REVIEW

**Comment by glennw** [comment]
This PR also includes #2552. It may be easier to review this PR after that lands.

The try run has several orange items, but I think they are all OK:

* Wr5 is 2x new PASS.
* R3 is 1x new PASS.
* R4 is 2x new PASS.
* rust item looks to be an unrelated build failure.
* mda2 looks like an unrelated intermittent.
* wpt2 looks like an unrelated intermittent.

@staktrace, does that look correct to you?

r? @mrobinson or @kvark

**Comment by glennw** [comment]
With try run link:

https://treeherder.mozilla.org/#/jobs?repo=try&revision=2b31e1906983c87cc73b8b9496d1de0036650eb9

**Comment by staktrace** [comment]
@glennw I agree with your try push assessment, I think this is fine to land.

**Comment by staktrace** [comment]
Surprise! The unexpected-pass results are all actually from turning off mipmaps, not from this PR. https://treeherder.mozilla.org/#/jobs?repo=try&revision=b114b7b9e9565b2f03470931c6574040e354aa10

**Comment by glennw** [comment]
@staktrace I thought that might actually be the case here :)

**Comment by glennw** [comment]
I've rebased this on top of master, now down to one commit with +269 / -410 lines :)

File: wrench/reftests/text/shadow-clip.yaml [hunk 1]
```diff
@@ -12,7 +12,7 @@ root:
         -
           type: "shadow"
           bounds: [0, 0, 200, 200]
-          blur-radius: 1
```
Comment (Gankro) [inline wrench/reftests/text/shadow-clip.yaml]: why was this necessary?

**Comment by glennw** [comment]
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[webrender/src/display_list_flattener.rs, line 941 at r1](https://reviewable.io:443/reviews/servo/webrender/2556#-L8YIXFykHqFgWgRx7Ct:-L8ZA99H0NFBLhBVjTPS:bbrp1og) ([raw file](https://github.com/servo/webrender/blob/5b8a81ebf1794e2f4b9bad1ede0eff3182987f9d/webrender/src/display_list_flattener.rs#L941)):*
<details><summary><i>Previously, Gankro (Alexis Beingessner) wrote…</i></summary><blockquote>

I don't recall us handling local clip sources around here before, why now?
</blockquote></details>

Because we now draw text decorations as a clip source - this has only landed recently.

---

*[webrender/src/display_list_flattener.rs, line 957 at r1](https://reviewable.io:443/reviews/servo/webrender/2556#-L8YJVvLsdspGX65os82:-L8ZAX4XneqlSu1olbsQ:bcs04fq) ([raw file](https://github.com/servo/webrender/blob/5b8a81ebf1794e2f4b9bad1ede0eff3182987f9d/webrender/src/display_list_flattener.rs#L957)):*
<details><summary><i>Previously, Gankro (Alexis Beingessner) wrote…</i></summary><blockquote>

Hmm, is it actually possible for a primitive to be in a non-visible container while the shadow isn't? (is a shadow ever in a different container from its contents?)

PrimitiveContainers are a new notion to me so idk what they represent (kvark says they're the new Picture, but I still see Pictures are alive and well?)

Is it accurate to say they're just temporary metadata used during flattening? so e.g. the fact that we do tons of non-recycled calls to create_shadow is "fine"? If so I'd prefer to call create_shadow something that indicates it's uh, softer, than it sounds. with_shadow?
</blockquote></details>

A PrimitiveContainer is just an enum used to handle pushing all of the primitive types. In this case, the is_visible just checks if alpha > 0, which is what we used to do manually in the add_text and add_line functions.

---

*[webrender/src/display_list_flattener.rs, line 1419 at r1](https://reviewable.io:443/reviews/servo/webrender/2556#-L8YM797-BIOQghBahgN:-L8ZAkqq4I3KOaKTFzuq:b-896fix) ([raw file](https://github.com/servo/webrender/blob/5b8a81ebf1794e2f4b9bad1ede0eff3182987f9d/webrender/src/display_list_flattener.rs#L1419)):*
<details><summary><i>Previously, Gankro (Alexis Beingessner) wrote…</i></summary><blockquote>

do -> to
</blockquote></details>

Done.

---

*[webrender/src/display_list_flattener.rs, line 1449 at r1](https://reviewable.io:443/reviews/servo/webrender/2556#-L8YLaWNO4vyY6vmsa-K:-L8ZApOTqlQiEjlT5Mlt:b-l9u3m3) ([raw file](https://github.com/servo/webrender/blob/5b8a81ebf1794e2f4b9bad1ede0eff3182987f9d/webrender/src/display_list_flattener.rs#L1449)):*
<details><summary><i>Previously, Gankro (Alexis Beingessner) wrote…</i></summary><blockquote>

😭
</blockquote></details>

Is that a good reaction, or an issue?

---

*[webrender/src/picture.rs, line 193 at r1](https://reviewable.io:443/reviews/servo/webrender/2556#-L8YZgpLy_LlP4XZaMMb:-L8ZAx9L2tBO23emmP30:b-vvdvja) ([raw file](https://github.com/servo/webrender/blob/c32585c1ed94d6cad938ecd0ef7db566a367a00d/webrender/src/picture.rs#L193)):*
<details><summary><i>Previously, Gankro (Alexis Beingessner) wrote…</i></summary><blockquote>

Why is FilterOp::Blur handling deferred to prepare_prim_for_render, but not e.g. DropShadow?
</blockquote></details>

Because the DropShadow code is much broken. I'm working on fixing the drop shadow code as a follow up patch.

---

*[webrender/src/tiling.rs, line 409 at r1](https://reviewable.io:443/reviews/servo/webrender/2556#-L8YFlkD7sigICPCkaFO:-L8ZB6O1rirRpjjnZwWQ:bb1h3wy) ([raw file](https://github.com/servo/webrender/blob/5b8a81ebf1794e2f4b9bad1ede0eff3182987f9d/webrender/src/tiling.rs#L409)):*
<details><summary><i>Previously, Gankro (Alexis Beingessner) wrote…</i></summary><blockquote>

Why isn't this just `if let PictureKind::Image { frame_output_pipeline_id: Some(pipeline_id), .. } = pic.kind { self.outputs.push(...) }`?
</blockquote></details>

I got a rustc compile error, since there's just the one PictureKind now (I'm intending to remove that enum as a follow up, but left it here for now to avoid a heap of indentation changes).

---

*[wrench/reftests/text/blurred-shadow-local-clip-rect-ref.png, line  at r1](https://reviewable.io:443/reviews/servo/webrender/2556#-L8YEmvB5vOyR1-fGHm8:-L8ZBKFXkYyWINnYB9vK:b6s3akd) ([raw file](https://github.com/servo/webrender/blob/5b8a81ebf1794e2f4b9bad1ede0eff3182987f9d/wrench/reftests/text/blurred-shadow-local-clip-rect-ref.png#L)):*
<details><summary><i>Previously, Gankro (Alexis Beingessner) wrote…</i></summary><blockquote>

All the content seems to be shifting down a ~pixel here. That seems suspicious?
</blockquote></details>

I didn't look into that diff too closely, since the try run passed. I can try to find the exact cause though, if you like?

---

*[wrench/reftests/text/shadow-clip.yaml, line 15 at r1](https://reviewable.io:443/reviews/servo/webrender/2556#-L8YEfKoelzV7IIvjiQj:-L8ZBWO5rdcOR0v5UpVK:blp49ur) ([raw file](https://github.com/servo/webrender/blob/5b8a81ebf1794e2f4b9bad1ede0eff3182987f9d/wrench/reftests/text/shadow-clip.yaml#L15)):*
<details><summary><i>Previously, Gankro (Alexis Beingessner) wrote…</i></summary><blockquote>

why was this necessary?
</blockquote></details>

A blur radius of 1 was producing a fractional offset in the primitive, due to the halving of the blur radius, and then scaling by 3 for the blur range. This was producing pixel perfect results on all the hardware I tested, but OSMesa seemed to be rounding it incorrectly, causing the test to fail. Since it only affected OSMesa, and the blur radius being 2 doesn't affect what this test is checking, I used a whole number to work around that for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2556)*

**Comment by Gankro** [comment]
Review status: 11 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[webrender/src/display_list_flattener.rs, line 1449 at r1](https://reviewable.io:443/reviews/servo/webrender/2556#-L8YLaWNO4vyY6vmsa-K:-L8ZZ78AMBB4VScvvzBH:b99sa82) ([raw file](https://github.com/servo/webrender/blob/5b8a81ebf1794e2f4b9bad1ede0eff3182987f9d/webrender/src/display_list_flattener.rs#L1449)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Is that a good reaction, or an issue?
</blockquote></details>

tears of joy!

---

*[webrender/src/tiling.rs, line 409 at r1](https://reviewable.io:443/reviews/servo/webrender/2556#-L8YFlkD7sigICPCkaFO:-L8ZYyhIgRB69dDHPW5B:bvxyozc) ([raw file](https://github.com/servo/webrender/blob/5b8a81ebf1794e2f4b9bad1ede0eff3182987f9d/webrender/src/tiling.rs#L409)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

I got a rustc compile error, since there's just the one PictureKind now (I'm intending to remove that enum as a follow up, but left it here for now to avoid a heap of indentation changes).
</blockquote></details>

Yes that's why I moved the `Some()` match inline, which makes the whole pattern fallible.

---

*[wrench/reftests/text/blurred-shadow-local-clip-rect-ref.png, line  at r1](https://reviewable.io:443/reviews/servo/webrender/2556#-L8YEmvB5vOyR1-fGHm8:-L8ZZJEUswATXRA6krSE:bdawlrb) ([raw file](https://github.com/servo/webrender/blob/5b8a81ebf1794e2f4b9bad1ede0eff3182987f9d/wrench/reftests/text/blurred-shadow-local-clip-rect-ref.png#L)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

I didn't look into that diff too closely, since the try run passed. I can try to find the exact cause though, if you like?
</blockquote></details>

It's also possible github is just being dumb here. But wait, how did you get a passing try *and* a diff?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2556)*

**Comment by glennw** [comment]
Review status: 10 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[webrender/src/tiling.rs, line 409 at r1](https://reviewable.io:443/reviews/servo/webrender/2556#-L8YFlkD7sigICPCkaFO:-L8Z_yM31ca0LxzSp-hC:b-bon5f2) ([raw file](https://github.com/servo/webrender/blob/5b8a81ebf1794e2f4b9bad1ede0eff3182987f9d/webrender/src/tiling.rs#L409)):*
<details><summary><i>Previously, Gankro (Alexis Beingessner) wrote…</i></summary><blockquote>

Yes that's why I moved the `Some()` match inline, which makes the whole pattern fallible.
</blockquote></details>

Ah, I misread your comment. Fixed now.

---

*[wrench/reftests/text/blurred-shadow-local-clip-rect-ref.png, line  at r1](https://reviewable.io:443/reviews/servo/webrender/2556#-L8YEmvB5vOyR1-fGHm8:-L8Za6Apt3zG4eN5y8fs:bmtvbsy) ([raw file](https://github.com/servo/webrender/blob/5b8a81ebf1794e2f4b9bad1ede0eff3182987f9d/wrench/reftests/text/blurred-shadow-local-clip-rect-ref.png#L)):*
<details><summary><i>Previously, Gankro (Alexis Beingessner) wrote…</i></summary><blockquote>

It's also possible github is just being dumb here. But wait, how did you get a passing try *and* a diff?
</blockquote></details>

Oh, I meant I hadn't looked too closely at the wrench reference image difference, since the gecko try had passed. I can take a closer look at this difference before we merge though, if you prefer, to see if it's actually a bug with the previous image or the new one?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2556)*

**Comment by Gankro** [comment]
Review status: 10 of 12 files reviewed at latest revision, 1 unresolved discussion.

---

*[wrench/reftests/text/blurred-shadow-local-clip-rect-ref.png, line  at r1](https://reviewable.io:443/reviews/servo/webrender/2556#-L8YEmvB5vOyR1-fGHm8:-L8Zc2uK7oNfsys14Ubk:b-97932c) ([raw file](https://github.com/servo/webrender/blob/5b8a81ebf1794e2f4b9bad1ede0eff3182987f9d/wrench/reftests/text/blurred-shadow-local-clip-rect-ref.png#L)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Oh, I meant I hadn't looked too closely at the wrench reference image difference, since the gecko try had passed. I can take a closer look at this difference before we merge though, if you prefer, to see if it's actually a bug with the previous image or the new one?
</blockquote></details>

I'd appreciate it, but not a blocker if you think you have better stuff to do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2556)*

**Comment by glennw** [comment]
I'll take a look today, and see what causes the difference. r=you if it turns out to be expected? (otherwise if there's a bug I'll put up an extra commit + gecko try).

**Comment by Gankro** [comment]
yep, r=me with the image investigate

**Comment by glennw** [comment]
@Gankro You're quite right - that reference image was incorrectly clipped. I guess we don't have Gecko test coverage for that. Pushed a commit that restores the local rect inflate for blurs, will kick off a gecko try shortly.

**Comment by glennw** [comment]
Try run here https://treeherder.mozilla.org/#/jobs?repo=try&revision=e5cabd0b6e5a2c58b75f4fd20d55a0c782166b97 has one new PASS and one new FAIL.

I'll investigate the failure today.

**Comment by glennw** [comment]
Hmmm, seems to not reproduce locally so far :disappointed:

**Comment by glennw** [comment]
Latest try run (not quite complete yet):

https://treeherder.mozilla.org/#/jobs?repo=try&revision=a51d30f477a9adbf31ba1e3053275b5177a0e06e

There is a new PASS in R7. The other orange results look unrelated, so I think this is ready to merge now.

**Comment by glennw** [comment]
Try run looks good now.

@bors-servo r=Gankro :rocket:
